README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ONEST

The Observers Needed to Evaluate Subjective Tests software implements a
statistical method in Reisenbichler et al. (2020[1]), to determine the
minimum number of evaluators needed to estimate agreement involving a
large number of raters. This method could be utilized by regulatory
agencies, such as the FDA, when evaluating agreement levels of a newly
proposed subjective laboratory test. Input to the program should be
binary(1/0) pathology data, where “0” may stand for negative and “1” for
positive. The example datasets in this software are from Rimm et
al. (2017[2]) (the SP142 assay), and Reisenbichler et al. 2020. This
program can run in R version 3.5.0 and above.

## Installation

You can install ONEST from github with:

``` r
# install.packages("devtools")
devtools::install_github("hangangtrue/ONEST")
```

You can also install the released version of ONEST from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("ONEST")
```

## Example

This is a basic example which shows you how this software works:

``` r
library(ONEST)

# load in the sp142_bin dataset
data('sp142_bin')
ONEST_main(sp142_bin)
```

![](man/figures/README-example-1.png)<!-- -->![](man/figures/README-example-2.png)<!-- -->![](man/figures/README-example-3.png)<!-- -->![](man/figures/README-example-4.png)<!-- -->![](man/figures/README-example-5.png)<!-- -->![](man/figures/README-example-6.png)<!-- -->

    #> $consistency
    #>       consist_p consist_low
    #>  [1,] 0.6911795   0.6427088
    #>  [2,] 0.5367693   0.4640632
    #>  [3,] 0.4595634   0.3747395
    #>  [4,] 0.4209597   0.3300768
    #>  [5,] 0.4016573   0.3077448
    #>  [6,] 0.3920057   0.2965783
    #>  [7,] 0.3871797   0.2909948
    #>  [8,] 0.3847665   0.2882029
    #>  [9,] 0.3835598   0.2868068
    #> [10,] 0.3829564   0.2861087
    #> [11,] 0.3826547   0.2857597
    #> [12,] 0.3825039   0.2855851
    #> [13,] 0.3824284   0.2854978
    #> [14,] 0.3823907   0.2854542
    #> [15,] 0.3823718   0.2854324
    #> [16,] 0.3823624   0.2854214
    #> [17,] 0.3823577   0.2854160
    #> 
    #> $difference
    #>        diff_consist    diff_high
    #>  [1,] -1.544102e-01 1.786456e-01
    #>  [2,] -7.720588e-02 8.932368e-02
    #>  [3,] -3.860371e-02 4.466273e-02
    #>  [4,] -1.930243e-02 2.233203e-02
    #>  [5,] -9.651598e-03 1.116646e-02
    #>  [6,] -4.826038e-03 5.583506e-03
    #>  [7,] -2.413163e-03 2.791919e-03
    #>  [8,] -1.206665e-03 1.396057e-03
    #>  [9,] -6.033806e-04 6.980838e-04
    #> [10,] -3.017172e-04 3.490731e-04
    #> [11,] -1.508736e-04 1.745539e-04
    #> [12,] -7.544503e-05 8.728646e-05
    #> [13,] -3.772701e-05 4.364843e-05
    #> [14,] -1.886594e-05 2.182703e-05
    #> [15,] -9.434279e-06 1.091503e-05
    #> [16,] -4.717841e-06 5.458327e-06
    #> 
    #> $estimates
    #>      size_case size_rater         p    p_plus   p_minus
    #> [1,]        68         18 0.4984245 0.2794118 0.1029412
    #> 
    #> $empirical
    #>       lower_bound      mean upper_bound
    #>  [1,]   0.6029412 0.7928382   0.9264706
    #>  [2,]   0.5294118 0.6938971   0.8529412
    #>  [3,]   0.4411765 0.6277941   0.8088235
    #>  [4,]   0.4117647 0.5815147   0.7352941
    #>  [5,]   0.3970588 0.5455735   0.6911765
    #>  [6,]   0.3823529 0.5155294   0.6470588
    #>  [7,]   0.3676471 0.4906471   0.6176471
    #>  [8,]   0.3676471 0.4701912   0.5882353
    #>  [9,]   0.3529412 0.4492500   0.5735294
    #> [10,]   0.3529412 0.4326765   0.5441176
    #> [11,]   0.3529412 0.4171176   0.5294118
    #> [12,]   0.3529412 0.4025882   0.5000000
    #> [13,]   0.3529412 0.3904706   0.4852941
    #> [14,]   0.3529412 0.3791176   0.4558824
    #> [15,]   0.3529412 0.3688971   0.4558824
    #> [16,]   0.3529412 0.3609559   0.4411765
    #> [17,]   0.3529412 0.3529412   0.3529412

Some more details and examples can be found in vignettes of the package.

[1] Reisenbichler, E. S., Han, G., Bellizzi, A., Bossuyt, V., Brock, J.,
Cole, K., Fadare, O., Hameed, O., Hanley, K., Harrison, B. T., Kuba, M.
G., Ly, A., Miller, D., Podoll, M., Roden, A. C., Singh, K., Sanders, M.
A., Wei, S., Wen, H., Pelekanou, V., Yaghoobi, V., Ahmed, F., Pusztai,
L., and Rimm, D. L. (2020) “Prospective multi-institutional evaluation
of pathologist assessment of PD-L1 assays for patient selection in
triple negative breast cancer,” Mod Pathol, DOI:
10.1038/s41379-020-0544-x; PMID: 32300181.

[2] Rimm, D. L., Han, G., Taube, J. M., Yi, E. S., Bridge, J. A.,
Flieder, D. B., Homer, R., West, W. W., Wu, H., Roden, A. C., Fujimoto,
J., Yu, H., Anders, R., Kowalewski, A., Rivard, C., Rehman, J.,
Batenchuk, C., Burns, V., Hirsch, F. R., and Wistuba,, II (2017) “A
Prospective, Multi-institutional, Pathologist-Based Assessment of 4
Immunohistochemistry Assays for PD-L1 Expression in Non-Small Cell Lung
Cancer,” JAMA Oncol, 3(8), 1051-1058, DOI: 10.1001/jamaoncol.2017.0013,
PMID: 28278348.


NEWS.md
# ONEST 0.1.0

* Added a `NEWS.md` file to track changes to the package.
* Added the 'ONEST_inflation_test' function.
* Added a new example data set 'NCCN_22c3_t.rdata'.


R/ONEST_inflation_test.R
#' The ONEST_inflation_test
#'
#' @description This function uses the binary (0/1) pathology data to test if the agreement proportion will converge to 0 with an increasing number of observers.
#'
#' @param Y pathology data as a matrix (missing values are allowed)
#'
#' @importFrom stats pchisq
#'
#' @usage ONEST_inflation_test(Y)
#' @return
#' p_value: chi-square p-value, a small p-value indicates significant evidence that the observers’ agreement will converge to a non-zero proportion.
#'
#' @export
#'
#' @examples
#' data("sp263_bin")
#' ONEST_inflation_test(sp263_bin)
ONEST_inflation_test=function(Y){
  # Step 1: estimate the parameter p under H0;
  # Initiate n, M, C0, C1;
  # Input Y;
  sum_1 = sum(Y,na.rm = T)
  sum_0 = sum(1-Y,na.rm = T)
  p = sum_1/(sum_1 + sum_0)
  n = nrow(Y)
  M = ncol(Y)

  # calculate theta_0 and theta_1;
  # theta_0: number of cases all being 0;
  # theta_1: number of cases all being 1;
  C0  = 0; CM = 0;
  for (i in 1:n){
    if (sum(Y[i,] == 1, na.rm = TRUE)==0){
      C0 = C0 + 1
    }

    if (sum(Y[i,] == 1, na.rm = TRUE) == sum(Y[i,],na.rm = TRUE) + sum(Y[i,] == 0, na.rm = TRUE)){
      CM = CM + 1
    }
  }
  if (C0>0){
    C0 = C0#-0.5
  }
  if (CM>0){
    CM = CM#-0.5
  }

  # theta_0
  # theta_1

  U = matrix(c(CM/(p^M)-n, C0/(1-p)^M-n),ncol = 2)
  I = matrix(c(n-n*p^M+n*p^M*((1-p^M)/p^M)^2,-3*n,-3*n,n-n*(1-p)^M+n*(1-p)^M*((1-(1-p)^M)/(1-p)^M)^2),ncol = 2)
  #I = I;

  # Test statistic;
  S = U%*%solve(I)%*%t(U)
  S2 = matrix(c(n,n),ncol = 2)%*%solve(I)%*%t(matrix(c(n,n),ncol = 2))

  # P-value;
  p1_value = pchisq(S,2,lower.tail = F)
  p2_value = pchisq(S2,2,lower.tail = F)

  # 3 by 3 covariance matrix, including p;
  U2 = matrix(c(U,0),ncol = 3)
  I2 = matrix(rep(0,9),ncol = 3)
  I2[1:2,1:2] = I
  I2[3,3] = M*n*(1/p + 1/(1-p))
  I2[1,3] = -1*(1/p+(M*n-1)*p^(M-1))
  I2[3,1] = I2[1,3]
  I2[2,3] = -1*(1/(1-p)+(M*n-1)*(1-p)^(M-1))
  I2[3,2] = I2[2,3]

  S3 = U2%*%solve(I2)%*%t(U2)
  p3_value = pchisq(S3,2,lower.tail = F)

  p_values = c(p1_value=p1_value,p2_value=p2_value,p3_value=p3_value)
  p_value=c(p_value=p3_value)
  return(p_value)

}


R/plotline.R
#' @title Plotline
#'
#' @description This function uses the binary (0/1) pathology data to plot the percentage of agreement.
#'
#' @param path pathology data as a matrix (missing values are allowed)
#' @param indi an indicator; if indi=0: then do not make plot;
#'             indi=1: then make plot, original scale.
#'             Labels on the x and y-axis are on the original scale
#' @param color the color used to draw pictures
#'
#' @importFrom graphics plot
#'
#' @usage plotline(path,indi=1,color="red")
#' @return
#' A matrix with percentage of agreement from the original scale
#'
#' @export
#'
#' @examples
#' data("sp142_bin")
#' plotline(sp142_bin,1,"red")
plotline = function(path, indi=1, color='red'){
  n = nrow(path)
  m = ncol(path)
  a = rep(0,m-1)

  # Concordance for original data
  concord  = matrix(a,ncol=1)
  M = m-1
  for (i in 1:M){
    temp = rep(0,n)
    I = i+1
    for (j in 1:n){
      comp = as.vector(unlist(path[j,1:I]))
      temp[j] = length(unique(comp))
    }
    concord[i] = sum(temp==1,na.rm = TRUE)/n
  }

  # Plot the data based on the indi status
  if (indi == 1){
    x_axis = 2:m
    plot(x_axis,concord, type = "o", col = color,lwd=1.8,xlim=c(2,m),ylim=c(0,1),
         xlab = 'Number of raters',ylab='Proportion of identical reads',main="Figure(1)")
  }

  C = concord
  return(C)
}


R/ONEST_main.R
#' @title ONEST main function
#'
#' @description This is the ONEST main function taking binary (0/1) pathology data as input. This function
#' utilizes the plotline function to produce the ONEST graphs and estimates.
#'
#' @param data a matrix containing the binary pathology data. Each row is the data from one case, and each column is the data from one rater. Missing values are allowed and can be denoted as NA or left blank. If there are n cases and k raters, the input ‘data’ is a matrix with dimension n by k.
#'
#' @importFrom graphics barplot
#' @importFrom graphics lines
#' @importFrom graphics legend
#' @importFrom stats quantile
#'
#' @usage ONEST_main(data)
#'
#' @return
#' consist_p: a vector of length k-1, indicating proportion of identical reads among a set of pathologists. For example, the first element of “consist_p” is the estimate of agreement percentage for 2 raters. The k-1 th element is the estimate of agreement percentage for k raters.
#'
#' consist_low: a vector of length k-1, indicating the lower bound of the agreement percentage with 95 percent confidence level corresponding to “consist_p”.
#'
#' diff_consist: a vector of length k-2, indicating the difference between the consist_p. For example, the first element of “diff_consist” is the estimated difference of agreement percentage after increasing from 2 to 3 raters. The k-2 th element is the difference of agreement percentage after increasing from k-1 to k raters.
#'
#' diff_high: a vector of length k-2, indicating the upper bound of the change of agreement percentage corresponding to “diff_consist” with 95 percent confidence level.
#'
#' size_case: number of cases n
#'
#' size_rater: number of raters k
#'
#' p: the probability of of being rated positive among the proportion of ‘1-p_plus-p_minus’ cases.
#'
#' p_plus: proportion of the cases rated positive by all raters.
#'
#' p_minus: proportion of the cases rated negative by all raters.
#'
#' empirical: a matrix of dimension k-1 by 3, including the empirical estimate of the agreement percentage, and the empirical 95 percent confidence intervals (CI) of the agreement percentage with equal tail probabilities on the two sides. The empirical estimate and CI were calculated by permuting the raters with 1000 random permutations, and using the mean, 2.5th percentile, and 97.5th percentile.
#'
#'
#' @export
#'
#' @examples
#' data("sp142_bin")
#' ONEST_main(sp142_bin)
ONEST_main = function(data){

  # Size of the data
  n = nrow(data)
  m = ncol(data)
  size_case = n
  size_rater = m
  M = m-1

  ## Step 1, plot the data
  # 1) Plot the agreement percentage in the order of columns in the input
  color = 'black'
  C = plotline(data,1,color)
  lines(0,0,main="Figure(1)")

  # 2) Plot 100 randomly chosen permutations
  concord = matrix(rep(0,100*M),ncol=100)
  for (i in 1:100) {
    concord[,i] = plotline(data[,sample(m)],0)
  }
  x_axis = 2:m
  plot(x_axis,concord[,100], type = "o", col = color,lwd=1.8,xlim=c(2,m),ylim=c(0,1),main="Figure(2)",
       xlab = 'Number of raters',ylab='Proportion of identical reads')
  for (i in 1:99){
    lines(x_axis,concord[,i], type = "o", col = color,lwd=1.8,xlim=c(2,m),ylim=c(0,1))
  }

  # 3) Plot the empirical confidence interval
  simN = 1000
  D = matrix(rep(0,simN*(m-1)),ncol=simN)
  for (i in 1:simN){
    D[,i]  = plotline(data[,sample(m)],0)
  }

  # Calculate mean, 5 th, and 95 th percentiles
  plot_data = matrix(rep(0,3*(m-1)),ncol=3)
  for (j in 1:M){
    plot_data[j,1] = mean(D[j,])
    plot_data[j,2] = quantile(D[j,],0.025,names=FALSE,type=1)
    plot_data[j,3] = quantile(D[j,],0.975,names=FALSE,type=1)
  }

  # plot the mean and empirical 95# CI
  x_axis = 2:m
  plot(x_axis,plot_data[,1], type = "o", col = color,lwd=1.8,xlim=c(2,m),ylim=c(0,1),
       xlab = 'Number of raters',ylab='Proportion of agreement',main="Figure(3)")
  lines(x_axis,plot_data[,2],lty=2, col = color,lwd=1.8,xlim=c(2,m),ylim=c(0,1))
  lines(x_axis,plot_data[,3],lty=2, col = color,lwd=1.8,xlim=c(2,m),ylim=c(0,1))
  legend("topright", c("Mean","95% empirical CI"),lty=1:2,pch = c(1,NA),cex = 0.6)

  # 4) barchart: the x axis is the
  # case number and the Y axis is the number of pathologists that called
  # that case positive, sorted from lowest to highest on the y axis.
  sum_D = matrix(rep(0,n),ncol=1)
  mean_D = matrix(rep(0,n),ncol=1)
  for(k in 1:n){
    sum_D[k] = sum(data[k,],na.rm = TRUE)
    mean_D[k] = sum(data[k,],na.rm = TRUE)/length(which(data[k,]>-1))
  }
  a = tabulate(sum_D)
  # figure(4)
  # Relative proportion
  barplot(sort(mean_D),xlab = 'Cases',ylab='Proportion of positive ratings',col='white',main="Figure(4)")

  ## Step 2, model the data;

  # Input data = sp142_bin;

  data_temp = data

  # Total number of observations;
  N = n*m
  # count the number of nan;
  nan_num = 0
  for (i in 1:n){
    for (j in 1:m){
      if (is.na(data_temp[i,j])==1){
        nan_num = nan_num + 1
      }
    }
  }

  # total number of positve reads;
  K = sum(data_temp==1,na.rm = TRUE)
  # calculate p_plus and p_minus, the proportions of all positive and negative;
  p_plus  = 0; p_minus = 0;
  for (i in 1:n){
    if (sum(data_temp[i,]==1,na.rm = TRUE)==0){
      p_minus = p_minus + 1/n
    }

    if (sum(data_temp[i,]==1,na.rm = TRUE)==sum(data_temp[i,],na.rm = TRUE)+sum(data_temp[i,]==0,na.rm = TRUE)){
      p_plus = p_plus + 1/n
    }
  }
  # [7/68 19/68]
  # [p_minus p_plus]
  # p_plus  = 19/68; p_minus = 7/68;

  # Calculate the probability 'p'
  prop = (K/(N-nan_num)-p_plus)/(1-p_plus-p_minus)

  # Calculate the consistancy probability trajectory;
  consist_p = rep(0,m-1)
  diff_consist = rep(0,m-2)
  o=m-2
  for (i in 1:M){
    consist_p[i] = (1-p_plus-p_minus)*(prop^(i+1)+(1-prop)^(i+1))+p_plus+p_minus
  }
  # Calculate the difference trajectory;
  for (i in 1:o){
    diff_consist[i] = consist_p[i+1]-consist_p[i]
  }

  ## Construct the upper bound for the difference using the lower bound for
  # p_plus+p_minus;
  p_c = p_plus+p_minus
  p_c_low = p_c - 1.645*sqrt(p_c*(1-p_c)/n)

  # Calculate the trajectory;
  consist_low = rep(0,m-1)
  diff_low = rep(0,m-2)
  for (i in 1:M){
    consist_low[i]=(1-p_c_low)*(prop^(i+1)+(1-prop)^(i+1))+p_c_low
  }

  for (i in 1:o){
    diff_low[i] = consist_low[i+1]-consist_low[i]
  }
  # figure(5)
  plot(x_axis,consist_p, type = "o", col = "red",lwd=1.8,ylim = c(0,1),pch=2,xlab = 'Number of raters',ylab='Proportion of identical reads',main="Figure(5)")
  lines(x_axis,consist_low,lty=2,lwd=1.8,col = "orange")
  # figure(6)
  plot(1:o,-diff_consist, type = "o", col = "red",lwd=1.8,pch=2,xlab = 'Number of raters',ylab='Difference between the consist_p',main="Figure(6)")
  lines(1:o,-diff_low,lty=2,lwd=1.8,col = "orange")

  diff_high = -diff_low
  p = prop
  lower_bound = plot_data[,2]
  upper_bound = plot_data[,3]
  mean = plot_data[,1]
  empirical = rbind(cbind(lower_bound,mean,upper_bound))
  consist = rbind(cbind(consist_p,consist_low))
  diff = rbind(cbind(diff_consist,diff_high))
  estimate = rbind(cbind(size_case,size_rater,p,p_plus,p_minus))
  results = list(consistency=consist,difference=diff,estimates=estimate,empirical=empirical)
  return(results)
}



R/ONEST_vignettes.R
#' @title ONEST vignettes function
#'
#' @description This function is only used to create vignettes. Although it can get the same results from ONEST_main(sp142_bin), it uses some precomputed data to decrease the time to get the results. Therefore, it can only be applied to the sp142_bin dataset. Please use ONEST_main instead.
#'
#' @param data a matrix containing the binary pathology data. Each row is the data from one case, and each column is the data from one rater. Missing values are allowed and can be denoted as NA or left blank. If there are n cases and k raters, the input ‘data’ is a matrix with dimension n by k.
#' @param empirical a matrix containing data used to plot the empirical confidence intervals for the sp142_bin dataset.
#'
#' @importFrom graphics barplot
#' @importFrom graphics lines
#' @importFrom graphics legend
#' @importFrom stats quantile
#'
#' @usage ONEST_vignettes(data,empirical)
#'
#' @return
#' consist_p: a vector of length k-1, indicating proportion of identical reads among a set of pathologists. For example, the first element of “consist_p” is the estimate of agreement percentage for 2 raters. The k-1 th element is the estimate of agreement percentage for k raters.
#'
#' consist_low: a vector of length k-1, indicating the lower bound of the agreement percentage with 95 percent confidence level corresponding to “consist_p”.
#'
#' diff_consist: a vector of length k-2, indicating the difference between the consist_p. For example, the first element of “diff_consist” is the estimated difference of agreement percentage after increasing from 2 to 3 raters. The k-2 th element is the difference of agreement percentage after increasing from k-1 to k raters.
#'
#' diff_high: a vector of length k-2, indicating the upper bound of the change of agreement percentage corresponding to “diff_consist” with 95 percent confidence level.
#'
#' size_case: number of cases n
#'
#' size_rater: number of raters k
#'
#' p: the probability of of being rated positive among the proportion of ‘1-p_plus-p_minus’ cases.
#'
#' p_plus: proportion of the cases rated positive by all raters.
#'
#' p_minus: proportion of the cases rated negative by all raters.
#'
#' empirical: a matrix of dimension k-1 by 3, including the empirical estimate of the agreement percentage, and the empirical 95 percent confidence intervals (CI) of the agreement percentage with equal tail probabilities on the two sides. The empirical estimate and CI were calculated by permuting the raters with 1000 random permutations, and using the mean, 2.5th percentile, and 97.5th percentile.
#'
#'
#' @export
#'
#' @examples
#' data('sp142_bin')
#' data('empirical')
#' ONEST_vignettes(sp142_bin,empirical)
ONEST_vignettes = function(data,empirical){

  # Size of the data
  n = nrow(data)
  m = ncol(data)
  size_case = n
  size_rater = m
  M = m-1

  ## Step 1, plot the data
  # 1) Plot the agreement percentage in the order of columns in the input
  color = 'black'
  C = plotline(data,1,color)
  lines(0,0,main="Figure(1)")

  # 2) Plot 100 randomly chosen permutations
  concord = empirical[,1:100]
  x_axis = 2:m
  plot(x_axis,concord[,100], type = "o", col = color,lwd=1.8,xlim=c(2,m),ylim=c(0,1),main="Figure(2)",
       xlab = 'Number of raters',ylab='Proportion of identical reads')
  for (i in 1:99){
    lines(x_axis,concord[,i], type = "o", col = color,lwd=1.8,xlim=c(2,m),ylim=c(0,1))
  }

  # 3) Plot the empirical confidence interval
  D=empirical
  # Calculate mean, 5 th, and 95 th percentiles
  plot_data = matrix(rep(0,3*(m-1)),ncol=3)
  for (j in 1:M){
    plot_data[j,1] = mean(D[j,])
    plot_data[j,2] = quantile(D[j,],0.025,names=FALSE,type=1)
    plot_data[j,3] = quantile(D[j,],0.975,names=FALSE,type=1)
  }

  # plot the mean and empirical 95# CI
  x_axis = 2:m
  plot(x_axis,plot_data[,1], type = "o", col = color,lwd=1.8,xlim=c(2,m),ylim=c(0,1),
       xlab = 'Number of raters',ylab='Proportion of agreement',main="Figure(3)")
  lines(x_axis,plot_data[,2],lty=2, col = color,lwd=1.8,xlim=c(2,m),ylim=c(0,1))
  lines(x_axis,plot_data[,3],lty=2, col = color,lwd=1.8,xlim=c(2,m),ylim=c(0,1))
  legend("topright", c("Mean","95% empirical CI"),lty=1:2,pch = c(1,NA),cex = 0.6)

  # 4) barchart: the x axis is the
  # case number and the Y axis is the number of pathologists that called
  # that case positive, sorted from lowest to highest on the y axis.
  sum_D = matrix(rep(0,n),ncol=1)
  mean_D = matrix(rep(0,n),ncol=1)
  for(k in 1:n){
    sum_D[k] = sum(data[k,],na.rm = TRUE)
    mean_D[k] = sum(data[k,],na.rm = TRUE)/length(which(data[k,]>-1))
  }
  a = tabulate(sum_D)
  # figure(4)
  # Relative proportion
  barplot(sort(mean_D),xlab = 'Cases',ylab='Proportion of positive ratings',col='white',main="Figure(4)")

  ## Step 2, model the data;

  # Input data = sp142_bin;

  data_temp = data

  # Total number of observations;
  N = n*m
  # count the number of nan;
  nan_num = 0
  for (i in 1:n){
    for (j in 1:m){
      if (is.na(data_temp[i,j])==1){
        nan_num = nan_num + 1
      }
    }
  }

  # total number of positve reads;
  K = sum(data_temp==1,na.rm = TRUE)
  # calculate p_plus and p_minus, the proportions of all positive and negative;
  p_plus  = 0; p_minus = 0;
  for (i in 1:n){
    if (sum(data_temp[i,]==1,na.rm = TRUE)==0){
      p_minus = p_minus + 1/n
    }

    if (sum(data_temp[i,]==1,na.rm = TRUE)==sum(data_temp[i,],na.rm = TRUE)+sum(data_temp[i,]==0,na.rm = TRUE)){
      p_plus = p_plus + 1/n
    }
  }
  # [7/68 19/68]
  # [p_minus p_plus]
  # p_plus  = 19/68; p_minus = 7/68;

  # Calculate the probability 'p'
  prop = (K/(N-nan_num)-p_plus)/(1-p_plus-p_minus)

  # Calculate the consistancy probability trajectory;
  consist_p = rep(0,m-1)
  diff_consist = rep(0,m-2)
  o=m-2
  for (i in 1:M){
    consist_p[i] = (1-p_plus-p_minus)*(prop^(i+1)+(1-prop)^(i+1))+p_plus+p_minus
  }
  # Calculate the difference trajectory;
  for (i in 1:o){
    diff_consist[i] = consist_p[i+1]-consist_p[i]
  }

  ## Construct the upper bound for the difference using the lower bound for
  # p_plus+p_minus;
  p_c = p_plus+p_minus
  p_c_low = p_c - 1.645*sqrt(p_c*(1-p_c)/n)

  # Calculate the trajectory;
  consist_low = rep(0,m-1)
  diff_low = rep(0,m-2)
  for (i in 1:M){
    consist_low[i]=(1-p_c_low)*(prop^(i+1)+(1-prop)^(i+1))+p_c_low
  }

  for (i in 1:o){
    diff_low[i] = consist_low[i+1]-consist_low[i]
  }
  # figure(5)
  plot(x_axis,consist_p, type = "o", col = "red",lwd=1.8,ylim = c(0,1),pch=2,xlab = 'Number of raters',ylab='Proportion of identical reads',main="Figure(5)")
  lines(x_axis,consist_low,lty=2,lwd=1.8,col = "orange")
  # figure(6)
  plot(1:o,-diff_consist, type = "o", col = "red",lwd=1.8,pch=2,xlab = 'Number of raters',ylab='Difference between the consist_p',main="Figure(6)")
  lines(1:o,-diff_low,lty=2,lwd=1.8,col = "orange")

  diff_high = -diff_low
  p = prop
  lower_bound = plot_data[,2]
  upper_bound = plot_data[,3]
  mean = plot_data[,1]
  empirical = rbind(cbind(lower_bound,mean,upper_bound))
  consist = rbind(cbind(consist_p,consist_low))
  diff = rbind(cbind(diff_consist,diff_high))
  estimate = rbind(cbind(size_case,size_rater,p,p_plus,p_minus))
  results = list(consistency=consist,difference=diff,estimates=estimate,empirical=empirical)
  return(results)
}



R/ONEST.R
#' @title The R ONEST package
#' @name ONEST
#' @docType package
#' @description The ONEST software implements a statistical method in Reisenbichler et al. (2020[1]), to determine the minimum number of evaluators needed to estimate agreement involving a large number of raters. This method could be utilized by regulatory agencies, such as the FDA, when evaluating agreement levels of a newly proposed subjective laboratory test. Input to the program should be binary(1/0) pathology data, where “0” may stand for negative and “1” for positive. The example datasets in this software are from Rimm et al. (2017[2]) (the SP142 assay), and Reisenbichler et al. 2020. This program can run in R version 3.5.0 and above.
#'
#'
#' @references
#' [1] Reisenbichler ES, Han G, Bellizzi A, Bossuyt V, Brock J, Cole K, et al. Prospective multi-institutional evaluation of pathologist assessment of PD-L1 assays for patient selection in triple negative breast cancer. Mod Pathol. 2020, DOI: 10.1038/s41379-020-0544-x; PMID: 32300181.
#'
#' [2] Rimm DL, Han G, Taube JM, Yi ES, Bridge JA, Flieder DB, et al. A Prospective, Multi-institutional, Pathologist-Based Assessment of 4 Immunohistochemistry Assays for PD-L1 Expression in Non-Small Cell Lung Cancer. JAMA Oncol. 2017;3(8):1051-8, DOI: 10.1001/jamaoncol.2017.0013, PMID: 28278348.
NULL

#' Example dataset1
#'
#' A pathology reads dataset of triple negative breast cancer in Reisenbichler et al. (2020). A value of 0 means negative and a value of 1 means positive.
#'
#' \itemize{
#'   \item columns : different raters
#'   \item rows : different cases
#' }
#' @docType data
#' @keywords datasets
#' @name sp142_bin
#' @usage data('sp142_bin')
#' @references
#' [1] Reisenbichler ES, Han G, Bellizzi A, Bossuyt V, Brock J, Cole K, et al. Prospective multi-institutional evaluation of pathologist assessment of PD-L1 assays for patient selection in triple negative breast cancer. Mod Pathol. 2020, DOI: 10.1038/s41379-020-0544-x; PMID: 32300181.
NULL

#' Example dataset2
#'
#' A pathology reads dataset of triple negative breast cancer in Reisenbichler et al. (2020). A value of 0 means negative and a value of 1 means positive.
#'
#' \itemize{
#'   \item columns : different raters
#'   \item rows : different cases
#' }
#' @docType data
#' @keywords datasets
#' @name sp263_bin
#' @usage data('sp263_bin')
#' @references
#' [1] Reisenbichler ES, Han G, Bellizzi A, Bossuyt V, Brock J, Cole K, et al. Prospective multi-institutional evaluation of pathologist assessment of PD-L1 assays for patient selection in triple negative breast cancer. Mod Pathol. 2020, DOI: 10.1038/s41379-020-0544-x; PMID: 32300181.
NULL

#' Example dataset3
#'
#' NCCN SP142 stromal data from Rimm et al. (2017) used to illustrate the ONEST method.
#'
#' \itemize{
#'   \item columns : different raters
#'   \item rows : different cases
#' }
#' @docType data
#' @keywords datasets
#' @name NCCN_sp142
#' @usage data('NCCN_sp142')
#' @references
#' [1] Rimm DL, Han G, Taube JM, Yi ES, Bridge JA, Flieder DB, et al. A Prospective, Multi-institutional, Pathologist-Based Assessment of 4 Immunohistochemistry Assays for PD-L1 Expression in Non-Small Cell Lung Cancer. JAMA Oncol. 2017;3(8):1051-8, DOI: 10.1001/jamaoncol.2017.0013, PMID: 28278348.
NULL

#' Example dataset4
#'
#' NCCN SP142 tumor data from Rimm et al. (2017) used to illustrate the ONEST method.
#'
#' \itemize{
#'   \item columns : different raters
#'   \item rows : different cases
#' }
#' @docType data
#' @keywords datasets
#' @name NCCN_sp142_t
#' @usage data('NCCN_sp142_t')
#' @references
#' [1] Rimm DL, Han G, Taube JM, Yi ES, Bridge JA, Flieder DB, et al. A Prospective, Multi-institutional, Pathologist-Based Assessment of 4 Immunohistochemistry Assays for PD-L1 Expression in Non-Small Cell Lung Cancer. JAMA Oncol. 2017;3(8):1051-8, DOI: 10.1001/jamaoncol.2017.0013, PMID: 28278348.
NULL

#' Example dataset5
#'
#' NCCN 22c3 tumor data from Rimm et al. (2017) used to illustrate the ONEST method.
#'
#' \itemize{
#'   \item columns : different raters
#'   \item rows : different cases
#' }
#' @docType data
#' @keywords datasets
#' @name NCCN_22c3_t
#' @usage data('NCCN_22c3_t')
#' @references
#' [1] Rimm DL, Han G, Taube JM, Yi ES, Bridge JA, Flieder DB, et al. A Prospective, Multi-institutional, Pathologist-Based Assessment of 4 Immunohistochemistry Assays for PD-L1 Expression in Non-Small Cell Lung Cancer. JAMA Oncol. 2017;3(8):1051-8, DOI: 10.1001/jamaoncol.2017.0013, PMID: 28278348.
NULL

#' Precomputed dataset
#'
#' A precomputed dataset only used by the ONEST_vignettes function to create the vignettes.
#'
#' \itemize{
#'   \item columns : different permutations
#'   \item rows : agreement percentage
#' }
#' @docType data
#' @keywords datasets
#' @name empirical
#' @usage data('empirical')
NULL

